### PR TITLE
feat: allow triggering historical scan from CI

### PR DESCRIPTION
**Pull request by kopecs**
This adds the required glue code to run an additional scan with the required flags in the instance where historical scanning is enabled via the app and the secrets product is enabled.

Also fixes a minor issue where having only secrets enabled with no rules would not result in a warning we displayed in the case where only sast was enabled but no rules were sent.

Test plan: ~~additional testing in `cli/tests/e2e-pro`; `make test` in `cli/`.~~ will be moved to Pro due to CI config. See prior commits for example test. Manually tested with the test from 6a872f9367a708960940cdae4ddcda669d0832a6 with local copy of the pro binary.

Holding off on a changelog entry since I don't believe some required app work is done.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -61,7 +61,7 @@ repos:
       - id: requirements-txt-fixer
       - id: trailing-whitespace
         args: [--markdown-linebreak-ext=md]
-        exclude: "snapshots"
+        exclude: 'snapshots|.*\.network'
```
Comment (akuhlens) [inline .pre-commit-config.yaml]: This should likely be in the summary or a separate PR.

File: cli/src/semgrep/commands/ci.py [hunk 2]
```diff
@@ -342,10 +342,9 @@ def ci(
                 )
                 progress_bar.update(products_task, completed=100)
 
-            if (
-                scan_handler.rules == '{"rules":[]}'
-                and scan_handler.enabled_products == ["sast"]
-            ):
+            if scan_handler.rules == '{"rules":[]}' and set(
+                scan_handler.enabled_products
+            ).issubset({"sast", "secrets"}):
```
Comment (akuhlens) [inline cli/src/semgrep/commands/ci.py:347]: This could also be a separate PR or another thing mentioned in the summary.

File: cli/src/semgrep/commands/ci.py [hunk 3]
```diff
@@ -412,11 +411,47 @@ def ci(
...
+        "timeout_threshold": timeout_threshold,
+        "skip_unknown_extensions": (not scan_unknown_extensions),
+        "allow_untrusted_validators": allow_untrusted_validators,
+        "optimizations": optimizations,
+        "baseline_commit": metadata.merge_base_ref,
+        "baseline_commit_is_mergebase": True,
+        "diff_depth": diff_depth,
+        "dump_contributions": True,
+    }
+
```
Comment (akuhlens) [inline cli/src/semgrep/commands/ci.py:456]: You might comment why you pulled this out.

File: cli/src/semgrep/core_runner.py [hunk 4]
```diff
@@ -721,11 +723,14 @@ def _run_rules_direct_to_semgrep_core_helper(
             if dump_command_for_core
             else tempfile.NamedTemporaryFile("w+", suffix=".json")
         )
-        target_file = exit_stack.enter_context(
-            (state.env.user_data_folder / "semgrep_targets.txt").open("w+")
-            if dump_command_for_core
-            else tempfile.NamedTemporaryFile("w+")
-        )
+        if target_mode_config.is_historical_scan:
+            pass
```
Comment (akuhlens) [inline cli/src/semgrep/core_runner.py]: Could you comment why this is valid?

File: cli/src/semgrep/scan_report.py [hunk 5]
```diff
@@ -3,9 +3,9 @@
 ##############################################################################
 # Helpers to run_scan.py to report scan status
 import sys
+from collections.abc import Sequence
```
Comment (akuhlens) [inline cli/src/semgrep/scan_report.py]: What is this change doing?

**Review by akuhlens — CHANGES_REQUESTED** [review]
Here is some initial feedback. I will give more later.

**Comment by akuhlens** [comment]
Can you double check that the test results are what you think they should be. I will give you more feedback tomorrow morning. I am running late for a date.

File: cli/src/semgrep/rule_match.py [hunk 6]
```diff
@@ -18,16 +18,20 @@
 from attrs import evolve
 from attrs import field
 from attrs import frozen
+from boltons.cacheutils import cachedmethod
```
Comment (akuhlens) [inline cli/src/semgrep/rule_match.py]: Can you comment on the motivation for the line cache? Mostly just because caching is natoriously hard to get right.

File: cli/src/semgrep/scan_report.py (hunk 5 above)
Comment (akuhlens) [reply cli/src/semgrep/scan_report.py]: I *think* one of the CI failures is related to this...

**Review by akuhlens — CHANGES_REQUESTED** [review]
The CLI tests failures seem pretty related to this PR, could you investigate those, address the feedback here you agree with, and resubmit?

**Comment by kopecs** [comment]
> The CLI tests failures seem pretty related to this PR, could you investigate those, address the feedback here you agree with, and resubmit?

Yeah; I think we can not actually test pro in e2e-pro here in CI since we don't appear to copy the pro binary as part of those artefacts in general. I think I have to split it out and move them there.

File: cli/src/semgrep/commands/ci.py (hunk 2 above)
Comment (kopecs) [reply cli/src/semgrep/commands/ci.py:347]: Sure, I was just trying to clean some minor things up as I was going. I'll add a note in the summary & a changelog entry.

File: cli/src/semgrep/scan_report.py (hunk 5 above)
Comment (kopecs) [reply cli/src/semgrep/scan_report.py]: huh. Not sure, I think this was some IDE autocompletion or similar. I'll fix.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (kopecs) [reply .pre-commit-config.yaml]: Removing since this is only relevant for the test I need to move to pro.

File: cli/src/semgrep/rule_match.py (hunk 6 above)
Comment (kopecs) [reply cli/src/semgrep/rule_match.py]: Good flag; I think we ought to just cache getting the contents from git blobs anyway, so I've just gone ahead and moved this and added a comment (actually more text in a string; love Python) there instead.

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -61,7 +61,7 @@ repos:
       - id: requirements-txt-fixer
       - id: trailing-whitespace
         args: [--markdown-linebreak-ext=md]
-        exclude: "snapshots"
+        exclude: 'snapshots'
```
Comment (akuhlens) [inline .pre-commit-config.yaml]:
```suggestion
        exclude: "snapshots"
```

**Review by akuhlens — APPROVED** [review]
🙈

File: .pre-commit-config.yaml (hunk 7 above)
Comment (kopecs) [reply .pre-commit-config.yaml]: society when yaml.
